src/url-loader/indirect-url-resolver.ts

```diff
@@ -0,0 +1,122 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {FileRelativeUrlBrand, PackageRelativeUrl, UrlResolver} from '../index';
+import {FileRelativeUrl, ResolvedUrl} from '../model/url';
+
+import {FsUrlResolver} from './fs-url-resolver';
+
+type RuntimeUrl = string&RuntimeUrlBrand;
+
+
+// Declare these as classes rather than interfaces so that the properties
+// can be private.
+declare class RuntimeUrlBrand extends FileRelativeUrlBrand {
+  private RuntimeUrlBrand: never;
+}
+
+/**
+ * A URL resolver for very large codebases where source files map in an
+ * arbitrary but predetermined fashion onto URL space.
+ *
+ * It also separates the root directory – the root of all source code that's
+ * legal to load – from the package directory, which is how the user refers to
+ * files on the CLI or the IDE.
+ */
+export class IndirectUrlResolver extends FsUrlResolver implements UrlResolver {
+  private readonly runtimeUrlToResolvedUrl:
+      ReadonlyMap<RuntimeUrl, ResolvedUrl>;
+  private readonly resolvedUrlToRuntimeUrl:
+      ReadonlyMap<ResolvedUrl, RuntimeUrl>;
+
+  /**
+   * @param rootPath All loadable source code must be a descendent of this
+   *     directory. Should be the same as FsUrlLoader.
+   * @param packagePath The base directory for package-relative paths. Usually
+   * the current working directory.
```

Comment:
Why would this be the cwd? I would think the resolver configuration would be independent of the process's cwd, especially for the case of a multi-package repo.

Reply:
Don't get super caught up on the term package in package relative path. Package relative urls are just paths that come in from the user without further context about how to resolve them. In a megamonorepo the user is often in the directory of one package within the larger repo. rootPath is the base of the repo, packagePath is then the cwd.

Or put another way, suppose you're on the donuts project, you do the following:

```
cd /monorepo/food/dessert/circular/donuts
polymer lint components/donut-conveyor/donut-conveyor.html
```

In this example `/monorepo` is the rootPath, `/monorepo/food/dessert/circular/donuts` is the packagePath, `components/donut-conveyor/donut-conveyor.html` is the package-relative path.